File: java/step7-strategy-pattern/src/main/java/badcode/HardcodedRace.java
package badcode;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class Cars implements NumberGenerator, NumberGenerator {
    private List<RaceCar> cars;

    public Cars(List<String> names) {
        this.cars = new ArrayList<>();
        for (String name : names) {
            this.cars.add(new RaceCar(name));
        }
    }

    // 😱 Random이 메서드 내부에 하드코딩 — 테스트에서 결과를 예측할 수 없음!
    @Override
    public void playRound(RandomNumberGenerator NumberGenerator) {
        for (RaceCar car : cars) {
            int value = NumberGenerator.generate();
            car.move(value);
        }
    }

    public List<String> findWinners() {
        int maxPosition = 0;
        for (RaceCar car : cars) {
            if (car.getPosition() > maxPosition) {
                maxPosition = car.getPosition();
            }
        }
        List<String> winners = new ArrayList<>();
        for (RaceCar car : cars) {
            if (car.getPosition() == maxPosition) {
                winners.add(car.getName());
            }
        }
        return winners;
    }
}

File: ts/step7-strategy-pattern/hardcodedRace.ts
// race.ts
import { RaceCar } from './raceCar';
import { MovingStrategy } from './movingSTrategy';

export class Race {
    private cars: RaceCar[];
    private movingStrategy: MovingStrategy;

    constructor(names: string[], movingStrategy: MovingStrategy) {
        this.cars = names.map(name => new RaceCar(name));
        this.movingStrategy = movingStrategy;
    }

    public playRound(): void {
        for (const car of this.cars) {
            if (this.movingStrategy()) {
                car.move(4); 
            }
        }
    }

    public findWinners(): string[] {
        const maxPosition = Math.max(...this.cars.map(c => c.getPosition()));
        return this.cars
            .filter(c => c.getPosition() === maxPosition)
            .map(c => c.getName());
    }

    public getCars(): readonly RaceCar[] {
        return this.cars;
    }
}